api/processors/auxil/algorithms.py

```diff
@@ -0,0 +1,171 @@
+# DO NOT USE IN PROD, WORK IN PROGRESS
+
+from collections.abc import Iterable
+from dataclasses import dataclass
+from datetime import datetime
+
+from django.utils import timezone
+
+from api.models import AgeRange, Group, Student, Teacher
+from api.models.auxil.constants import (
+    DEFAULT_LESSON_DURATION_MIN,
+    MAX_AGE_KIDS_GROUP,
+    MAX_AGE_TEEN_GROUP,
+)
+from api.models.auxil.status_setter import StatusSetter
+from api.models.choices.communication_language_mode import CommunicationLanguageMode
+from api.models.choices.log_event_type import (
+    GroupLogEventType,
+    StudentLogEventType,
+    TeacherLogEventType,
+)
+from api.models.choices.status import (
+    GroupProjectStatus,
+    StudentSituationalStatus,
+    TeacherProjectStatus,
+    TeacherSituationalStatus,
+)
+from api.models.language_and_level import LanguageAndLevel
+from api.processors.auxil.log_event_creator import GroupLogEventCreator
+
+
+class GroupBuilderAlgorithm:
+    @dataclass
+    class GroupSizeRestriction:
+        # both min and max student numbers are inclusive:
+        # group size must fit into [min; max]
+        min: int
+        max: int
+
+        def validate_size(self, size: int) -> None:
+            if self.min > size:
+                raise ValueError(
+                    f"Group candidate must respect upper group boundaries, "
+                    f"but {size} > {self.max}"
+                )
+            if self.max < size:
+                raise ValueError(
+                    f"Group candidate must respect upper group boundaries, "
+                    f"but {size} > {self.max}"
+                )
+
+    @dataclass
+    class GroupCandidate:
+        language_and_level: LanguageAndLevel
+        communication_language_mode: CommunicationLanguageMode
+        age_range: AgeRange
+        teacher: Teacher
+        students: list[Student]
+        monday: datetime | None = None
+        tuesday: datetime | None = None
+        wednesday: datetime | None = None
+        thursday: datetime | None = None
+        friday: datetime | None = None
+        saturday: datetime | None = None
+        sunday: datetime | None = None
+
+        def __post_init__(self) -> None:
+            # Validate group size against age-related restrictions
+            restrictions = GroupBuilderAlgorithm._get_allowed_group_size(self.age_range)
+            restrictions.validate_size(len(self.students))
+
+    @staticmethod
+    def get_available_teachers() -> Iterable[Teacher]:
+        return Teacher.objects.filter(
+            project_status=TeacherProjectStatus.NO_GROUP_YET,
+            situational_status="",
```

Comment:
so yeah, unfortunately it gets a bit more complicated here after it got simpler there (i.e. in `project_status`) 😄 We have to include teachers that have `NO_GROUP_YET` and `WORKING` teachers whose `.can_take_more_groups` is `True`. You can check `.can_take_more_groups` for **both** `NO_GROUP_YET` and `WORKING` because we will set `simultaneous_groups` to `0` for teachers that don't want any regular groups and only want to host speaking clubs. So they might have `project_status=NO_GROUP_YET` but they won't be expecting to get any offers for regular groups at all.

Reply:
Another thing: how about creating a property `.is_available` in `Teacher` model? Then you can filter by it and get rid of the `is_teacher_available` static method completely.

Reply:
Thanks for both comments. Sounds very fair, will do

Reply:
Actually `can_take_more_groups` is the same as proposed `is_available`, as far as I understand.d